### PR TITLE
WIP: add a new segmentation module

### PR DESCRIPTION
**Pull request by poilvert**
@npinto,

First draft for a new segmentation module. So far I needed the `pad2d` function from
sthor so I added it to the module, even though in the future we will remove that extra
code and depend on sthor.

I also made a proxy to the Adjusted Rand Index metric from sklearn that seems fast
enough. That metric though, is only usable on _segmented images_ (on needs multiple
labels for each connected component).

I added a quick build script to build a small Cython module that I use in the warping error
code.

Finally here are the steps to make it (hopefully) work :
1. build the Cython extension module by executing `build_unique.sh`
2. run the demo code `warping_error_demo.py`

Cheese,

N.

### CODE REVIEW

File: bangmetric/segmentation.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        # -- find simple points in current annotation map that are also within
+        # the mask
+        simple = _compute_simple_points(out, mask)
+
+        # -- difference array between the current annotation map and the
+        # predicted map
+        diff = np.ma.array(np.abs(pred - out), mask=-simple)
+
+        # -- finding the largest deviation from the difference map
+        max_idx = np.unravel_index(diff.argmax(), diff.shape)
```
Comment (npinto) [inline bangmetric/segmentation.py]: Nice! I learned something here, I missed `unravel_index` from numpy >= 1.6.0 !

File: bangmetric/segmentation.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    bg_topological_numbers = _get_topo(bg_matrix, 0)
+
+    # -- building matrices of topological numbers
+    fg_topo_matrix = np.zeros(arr.shape, dtype=arr.dtype)
+    bg_topo_matrix = np.zeros(arr.shape, dtype=arr.dtype)
+
+    fg_topo_matrix[mask] = fg_topological_numbers
+    bg_topo_matrix[mask] = bg_topological_numbers
+
+    return (fg_topo_matrix == 1) * (bg_topo_matrix == 1)
```
Comment (npinto) [inline bangmetric/segmentation.py:245]: Possibility of having some topological breakage here.

File: bangmetric/segmentation.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+
+def _pad2d(arr_in, filter_shape2d, constant=-1, reverse_padding=False):
```
Comment (npinto) [inline bangmetric/segmentation.py]: Maybe in `utils`?

File: bangmetric/setup.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from distutils.core import setup
```
Comment (npinto) [inline bangmetric/setup.py]: Move this file to a more self-contained setup.py (see skimage/sthor example).

File: bangmetric/unique.pyx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import numpy as np
+cimport numpy as np
+
+from cython.parallel import prange
+
+DTYPE = np.int32
+ctypedef np.int32_t DTYPE_t
+
+
+def unique_cython(
```
Comment (npinto) [inline bangmetric/unique.pyx:14]: `unique*` may not be the best name here, what do you think?

File: bangmetric/warping_error_demo.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
```
Comment (npinto) [inline bangmetric/warping_error_demo.py]: Move to `examples`?

**Comment by npinto** [comment]
Looks awesome! I think we are missing the following
- clean cython build (see https://github.com/npinto/sthor/blob/develop/sthor/operation/setup.py#L7)
- smoke tests (Grab'ed)
- tests with random and pseudo-random images
- more complete tests based on Fiji (Java implementation).
- demo in `examples`
- (speed) benchmarks (if needed)
- topology-breaking points (ENH)

Anything else?

File: bangmetric/segmentation.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+
+    assert image_true.shape == image_pred.shape
+    assert image_true.ndim == 2
+    assert image_true.dtype in [np.bool, np.int]
+    assert image_pred.dtype in [np.float]
+
+    image_size = np.float(image_true.size)
+
+    # -- we transform the annotation image into an integer image
+    image_true = np.where(image_true > 0, 1, 0).astype(DTYPE_I)
```
Comment (npinto) [inline bangmetric/segmentation.py]: image_true = (image_true > 0).astype(...)

File: bangmetric/segmentation.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+
+        # -- finding the largest deviation from the difference map
+        max_idx = np.unravel_index(diff.argmax(), diff.shape)
+
+        # -- "flipping" the simple point with the largest deviation only if the
+        # deviation exceeds a threshold
+        if out[max_idx] > thr:
+            out[max_idx] = 1 - out[max_idx]
+        else:
+            return out
```
Comment (npinto) [inline bangmetric/segmentation.py]: make sure we don't end up in an infinite loop? (it looks fine here, but let's put a few watchdog / tests to make sure)

File: bangmetric/segmentation.py [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+    # -- then we compute a simple Hamming distance between the "best" warping
+    # and the ground truth
+    diff = best_warping - image_true
+    hamming = 1. / (image_size) * (np.linalg.norm(diff, ord='fro')) ** 2
+
+    return hamming
+
+
+def _warp_2d(true, pred, mask, thr=0.5):
```
Comment (npinto) [inline bangmetric/segmentation.py]: public function ? (for gradient descent)

Comment (npinto) [inline bangmetric/segmentation.py]: rename to `best_warp_*`, maybe with a kwarg `greedy`

**Comment by poilvert** [comment]
@npinto 

see if that fits you. I have some examples also in `examples`

**Comment by npinto** [comment]
Up ?

File: bangmetric/examples/Verena_metrics.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+from bangmetric.util import pad2d
+from bangmetric.precision_recall import average_precision
+from bangmetric.correlation import pearson, spearman
+from bangmetric.segmentation import warping_error_2d
+
+
+# -- parameters
+
+DIRNAME = './verenaImages'
+CONNECTOMICS_DIRNAME = '/home/poilvert/Dropbox/0__Shared_with_Poilvert/' + \
```
Comment (npinto) [inline bangmetric/examples/Verena_metrics.py:40]: hardcoded path again here...

File: bangmetric/examples/images.pkl [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
...x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x01\x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x01\x01\x00\x01\x01\x01\x01\x01\x01\x01\x01\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x01\x01\x00\x01\x01\x01\x01\x01\x01\x01\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x01\x01\x00\x01\x01\x01\x01\x01\x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x01\x01\x00\x01\x01\x01\x01\x01\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x01\x00\x01\x01\x00\x00\x00\x00\x00\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x01'
```
Comment (npinto) [inline bangmetric/examples/images.pkl:55]: do we need this here? if so, why?

File: bangmetric/segmentation.py [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+from scipy.ndimage import measurements
+from unique import unique_cython as _get_topo
+from util import pad2d
+import numpy as np
+
+DTYPE_I = np.int32
+DTYPE_F = np.float32
+
+
+def adjusted_rand_index(labels_true, labels_pred):
```
Comment (npinto) [inline bangmetric/segmentation.py:23]: not needed here, just use sklearn

File: bangmetric/examples/Verena_metrics.py (hunk 10 above)
Comment (npinto) [inline bangmetric/examples/Verena_metrics.py:40]: demo for ha (please discard)

`python code`
